torch/distributions/constraints.py

```diff
@@ -275,6 +276,17 @@ def check(self, value):
         return lower_triangular & positive_diagonal
 
 
+class _CorrCholesky(Constraint):
+    """
+    Constrain to lower-triangular square matrices with positive diagonals and each
+    row vector being of unit length.
+    """
+    def check(self, value):
+        row_norm = torch.linalg.norm(value, dim=-1)
+        unit_row_norm = ((row_norm <= 1.) & (row_norm >= (1 - 1e-6))).all(dim=-1)
```

Comment:
I would expect this to fail for float32 and matrices of dim > 100. If you want to be safer you could try something like:
```py
tol = torch.finfo(value.dtype).eps * value.size(-1) * 10  # 10 is an arbitrary fudge factor
row_norm = torch.linalg.norm(value.detach(), dim=-1)
return (row_norm - 1).abs().le(tol).all(dim=-1)
```

Reply:
Sound advice! Thanks, will make the change.

Reply:
Done. One thing that I realized was that as the size of the matrix grows, the transformed matrix returned by `CorrCholeskyTransform` will have 0 diagonal entries and will violate the lower cholesky constraint. I am not entirely sure what to do about it since numerically this is bound to happen. cc @fehiepsi. 

```python
>>> y = CorrCholeskyTransform()(torch.randn(19900))  # 200 x 200 matrix
>>> y  # last rows all have 0 diagonal entries

tensor([[ 1.0000,  0.0000,  0.0000,  ...,  0.0000,  0.0000,  0.0000],
        [-0.6596,  0.7516,  0.0000,  ...,  0.0000,  0.0000,  0.0000],
        [ 0.3552,  0.4238,  0.8332,  ...,  0.0000,  0.0000,  0.0000],
        ...,
        [ 0.8225, -0.5501, -0.1322,  ...,  0.0000,  0.0000,  0.0000],
        [ 0.5494, -0.6964, -0.3553,  ...,  0.0000,  0.0000,  0.0000],
        [ 0.1776, -0.4688,  0.7778,  ..., -0.0000,  0.0000,  0.0000]])
>>> lower_cholesky.check(y)

tensor(False)

>>> y @ y.T  # return a random correlation matrix

tensor([[ 1.0000, -0.6596,  0.3552,  ...,  0.8225,  0.5494,  0.1776],
        [-0.6596,  1.0000,  0.0842,  ..., -0.9560, -0.8859, -0.4695],
        [ 0.3552,  0.0842,  1.0000,  ..., -0.0512, -0.3961,  0.5125],
        ...,
        [ 0.8225, -0.9560, -0.0512,  ...,  1.0000,  0.8932,  0.3093],
        [ 0.5494, -0.8859, -0.3961,  ...,  0.8932,  1.0000,  0.1874],
        [ 0.1776, -0.4695,  0.5125,  ...,  0.3093,  0.1874,  1.0000]])
```

Reply:
In any case, I think we should relax the constraint to >= 0 for diagonal entries to allow for correlation matrices that are positive semi definite.

Reply:
I think we can replace
```
z1m_cumprod_sqrt = (1 - z).cumprod(-1).sqrt()
```
by
```
z1m_cumprod_sqrt = (1 - z).sqrt().cumprod(-1)
```
to improve precision but I am not sure if it will be stable enough. We might need to use float64 for a moderate dimension.

Reply:
It seems to really help for matrices of up to 200 dimensions (diag entries can be as small as 1e-38). For large matrices, using float64 would be a better idea. :D

Reply:
Good idea, that should help. I'll make the change. If we have issues with larger correlation matrices, we can discuss work-arounds.